include/sparrow/variable_size_binary_layout.hpp

```diff
@@ -0,0 +1,307 @@
+            is_const, typename L::const_data_iterator, typename L::data_iterator
+        >;
+
+        vs_binary_value_iterator() noexcept = default;
+        vs_binary_value_iterator(
+            offset_iterator offset_it,
+            data_iterator data_begin
+        );
+
+    private:
+
+        reference dereference() const;
+        void increment();
+        void decrement();
+        void advance(difference_type n);
+        difference_type distance_to(const self_type& rhs) const;
+        bool equal(const self_type& rhs) const;
+        bool less_than(const self_type& rhs) const;
+
+        offset_iterator m_offset_it;
+        const data_iterator m_data_begin;
+
+        friend class iterator_access;
+    };
+
+    /*
+     * @class variabe_size_binary_layout
+     *
+     * @brief Layout for arrays containing values consisting of a variable number of bytes.
+     *
+     * This layout is used to retrieve data in an array of values of a variable number of bytes
+     * (typically string objects). Values are stored contiguously in a data buffer, a single
+     * value is retrieved via an additional offset buffer, where each element is the beginning
+     * of the corresponinding value in the data buffer.
+     *
+     * Example:
+     *
+     * Let's consider the array ['please', 'allow', 'me', 'to', 'introduce', 'myself'].
+     * The internal buffers will be:
+     * - offset: [0, 6, 11, 13, 15, 24, 30]
+     * - data: ['p','l','e','a','s','e','a','l','l','o','w','m','e','t','o','i','n','t','r','o','d','u','c','e','m','y','s','e','l','f']
+     * 
+     * @tparam T the type of the data
+     * @tparam R the reference type to the data. This type is different form the reference type of the layout,
+     * which behaves like std::optional<R>.
+     * @tparam CR the const reference type to the data. This type is different from the const reference of the layout,
+     * which behaves like std::optional<CR>.
+     * @tparam OT type of the offset values. Must be std::int64_t or std::int32_t
+     */
+    template <class T, class R, class CR, class OT = std::int64_t>
+    class variable_size_binary_layout
+    {
+        static_assert(std::same_as<OT, std::int32_t> || std::same_as<OT, std::int64_t>);
+
+    public:
+
+        using self_type = variable_size_binary_layout<T, R, CR, OT>;
+        using inner_value_type = T;
+        using inner_reference = R;
+        using inner_const_reference = CR;
+        using value_type = std::optional<inner_value_type>;
+        using const_reference = const_reference_proxy<self_type>;
+        using size_type = std::size_t;
+
+        using const_bitmap_iterator = array_data::bitmap_type::const_iterator;
+        using const_value_iterator = vs_binary_value_iterator<self_type, true>;
+
+        explicit variable_size_binary_layout(array_data data);
+
+        size_type size() const;
+        const_reference operator[](size_type i) const;
+
+        const_value_iterator value_cbegin() const;
```

Comment:
Instead of pairs of functions here, I would prefer accessors, or view accessors if it's best.

btw what is the difference between `value_cbegin()` and `begin()`?

Reply:
`value_cbegin` iterates over the data buffer, while `begin` iterates over the reference_proxy. This allows some optimizations or element wise operations when you know you don't have any missing value (because null_count is 0), or when you don't care. Notice that we only provide const iterators on the values and the bitmap to prevent inconsistency (setting a value to something while not setting a corresponding bit to true for instance).

Not sure we want accessor instead, this is really meant to speed up iterations. For value / has_value access, ref proxies are enough.

Reply:
I meant something like:
```
auto values() const; // returns a range-view using the `const_value_iterator`
```

just grouping both iterator into a view if you prefer.

Reply:
Note that if `values()` returns a range of the actual buffer, therefore with element being `char`, then I would prefer it to have another name, like `raw_data()` or something in that direction.

Reply:
`values` returns a range on string_view, not on char.